### PR TITLE
First version of kpt live migrate command

### PR DESCRIPTION
**Pull request by seans3**
* First version of `kpt live migrate` command, that is not hooked in yet. Therefore, it will not affect the `kpt` binary yet.
* This command moves the inventory from a `ConfigMap` to a `ResourceGroup`
* Adds unit tests
* Also tested manually
```
$ kpt live migrate ~/testdata/testdata-namespace-1
inventory object migration...
  applying ResourceGroup CRD to the cluster...success
  updating Kptfile inventory values...success
  retrieve the current ConfigMap inventory...success (4 inventory objects)
  migrate inventory to ResourceGroup...success
  deleting old ConfigMap inventory object...success
inventory object migration...success

+ kubectl get po -n test-namespace
NAME    READY   STATUS    RESTARTS   AGE
pod-a   1/1     Running   0          2m3s
pod-b   1/1     Running   0          2m3s
pod-c   1/1     Running   0          2m3s
+ echo

+ kubectl get cm -n test-namespace
No resources found in test-namespace namespace.
+ echo

+ kubectl get resourcegroups.kpt.dev -o yaml
apiVersion: v1
items:
- apiVersion: kpt.dev/v1alpha1
  kind: ResourceGroup
  metadata:
    creationTimestamp: "2020-10-29T07:06:02Z"
    generation: 1
    labels:
      cli-utils.sigs.k8s.io/inventory-id: 0c26c31c1b6d5419d04e8986a3bc9853c377348d-1603955162291030527
    name: inventory-08877603
    namespace: default
    resourceVersion: "131381"
    selfLink: /apis/kpt.dev/v1alpha1/namespaces/default/resourcegroups/inventory-08877603
    uid: f893fadd-57ed-4db1-86ac-60dd141a24a9
  spec:
    resources:
    - group: ""
      kind: Pod
      name: pod-a
      namespace: test-namespace
    - group: ""
      kind: Pod
      name: pod-b
      namespace: test-namespace
    - group: ""
      kind: Pod
      name: pod-c
      namespace: test-namespace
    - group: ""
      kind: Namespace
      name: test-namespace
      namespace: ""
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
+ echo
```

### CODE REVIEW

File: commands/migratecmd.go [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+	mo := NewMigrateOptions(f, ioStreams)
+	cmd := &cobra.Command{
+		Use:                   "migrate DIRECTORY",
+		DisableFlagsInUseLine: true,
+		Short:                 i18n.T("Migrate inventory from ConfigMap to ResourceGroup custom resource"),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return mo.Run(cmd.InOrStdin(), args)
+		},
+	}
+	cmd.Flags().StringVar(&mo.initOptions.name, "name", "", "Inventory object name")
```
Comment (mortent) [inline commands/migratecmd.go:201]: I think we are trying avoid this pattern. Can we just us the runner pattern here like we do most other places and then only bind cobra flags to the runner?

File: commands/migratecmd.go [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+
+// NewCmdMigrate returns the cobra command for the migrate command.
+func NewCmdMigrate(f cmdutil.Factory, ioStreams genericclioptions.IOStreams) *cobra.Command {
+	mo := NewMigrateOptions(f, ioStreams)
+	cmd := &cobra.Command{
+		Use:                   "migrate DIRECTORY",
+		DisableFlagsInUseLine: true,
+		Short:                 i18n.T("Migrate inventory from ConfigMap to ResourceGroup custom resource"),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return mo.Run(cmd.InOrStdin(), args)
```
Comment (mortent) [inline commands/migratecmd.go:198]: So we are using both ioStreams and the streams from the cobra command. I think we should try to stick with one of them.

File: commands/migratecmd.go [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+	cmObjs, err := mo.retrieveInvObjs(cmInvInfo)
+	if err != nil {
+		return err
+	}
+	// Migrate the ConfigMap inventory objects to a ResourceGroup custom resource.
+	if err = mo.migrateObjs(cmObjs, bytes.NewReader(stdinBytes), args); err != nil {
+		return err
+	}
+	// Delete the old ConfigMap inventory object.
+	if err = mo.deleteConfigMapInv(cmInvInfo); err != nil {
```
Comment (phanimarupaka) [inline commands/migratecmd.go:76]: Why can't we just do `return mo.deleteConfigMapInv(cmInvInfo)` here

File: commands/migratecmd.go [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+		return err
+	}
+	// TODO(seans3): Remove the previous ConfigMap inventory template, if necessary.
+	return nil
+}
+
+// updateKptfile installs the "inventory" fields in the Kptfile.
+func (mo *MigrateOptions) updateKptfile(args []string) error {
+	fmt.Fprintf(mo.ioStreams.Out, "updating Kptfile values...")
```
Comment (phanimarupaka) [inline commands/migratecmd.go:85]: fmt.Fprint ?

File: commands/migratecmd.go [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+	if !mo.dryRun {
+		if err := mo.initOptions.Run(args); err != nil {
+			return err
+		}
+	}
+	fmt.Fprintf(mo.ioStreams.Out, "success\n")
```
Comment (phanimarupaka) [inline commands/migratecmd.go:91]: fmt.Fprint ?

File: commands/migratecmd.go [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+	}
+	fmt.Fprint(mr.ioStreams.Out, "success\n")
+	return nil
+}
+
+// stringToUnstructured transforms a single resource represented by
+// the passed string into a pointer to an "Unstructured" object,
+// or an error if one occurred.
+func stringToUnstructured(str string) (*unstructured.Unstructured, error) {
+	r := kio.ByteReader{OmitReaderAnnotations: true}
```
[the remark this answers is not in the document]
Comment (seans3) [reply commands/migratecmd.go:189]: Nice.

File: .golangci.yml [hunk 7]
```diff
@@ -45,7 +45,7 @@ linters:
 
 issues:
   exclude:
-     - Using the variable on range scope `tc` in function literal
+     - Using the variable on range scope .* in function literal
```
[the remark this answers is not in the document]
Comment (seans3) [reply .golangci.yml:48]: For some reason, using the older exclude statement with `tc`, I'm getting the following lint errors:
```
$ make lint
(which golangci-lint || go get github.com/golangci/golangci-lint/cmd/golangci-lint@v1.31.0)
/home/sean/go/bin/golangci-lint
/home/sean/go/bin/golangci-lint run ./...
commands/migratecmd_test.go:330:21: Using the variable on range scope `testing` in function literal (scopelint)
		t.Run(tn, func(t *testing.T) {
		                  ^
commands/migratecmd_test.go:331:19: Using the variable on range scope `findResourceGroupInv` in function literal (scopelint)
			actual, err := findResourceGroupInv(tc.objs)
			               ^
commands/migratecmd_test.go:332:10: Using the variable on range scope `isError` in function literal (scopelint)
			if tc.isError {
			      ^
commands/migratecmd_test.go:333:15: Using the variable on range scope `nil` in function literal (scopelint)
				if err == nil {
				          ^
commands/migratecmd_test.go:334:8: Using the variable on range scope `Fatalf` in function literal (scopelint)
					t.Fatalf("expected error but received none")
					  ^
commands/migratecmd_test.go:338:4: Using the variable on range scope `assert` in function literal (scopelint)
			assert.NoError(t, err)
			^
commands/migratecmd_test.go:339:10: Using the variable on range scope `expected` in function literal (scopelint)
			if tc.expected != actual {
			      ^
commands/migratecmd_test.go:340:7: Using the variable on range scope `Errorf` in function literal (scopelint)
				t.Errorf("expected ResourceGroup (%#v), got (%#v)", tc.expected, actual)
				  ^
make: *** [Makefile:61: lint] Error 1
```

I had to change the expression to `.*` to get rid of the lint errors. I still can't figure out why this is happening in only one test, when that same test has the same syntax as our other tests. Any help appreciated.

[the remark this answers is not in the document]
Comment (mortent) [reply .golangci.yml:48]: I also don't understand why the linter complains about that particular test function. It looks fine as far as I can tell. But just add `// nolint:scopelint` on the line before the function to suppress the errors for just this function.

[the remark this answers is not in the document]
Comment (seans3) [reply .golangci.yml]: Fixed.

[the remark this answers is not in the document]
Comment (seans3) [reply .golangci.yml]: I rearranged the tests so the final test is next to last. No other changes--strange, but it works. When I copied another test to the end of the file and renamed it, I got the same linting problems. Something about the the end of the file formatting was causing a problem I guess.

**Comment by mikebz** [comment]
General question and this might be on the wrong PR but is the a RG for every Kpt package?

**Comment by seans3** [comment]
> General question and this might be on the wrong PR but is the a RG for every Kpt package?

Every applied package has an inventory object. The inventory object has object references for objects within the applied package. That inventory object is currently a ConfigMap. We will move to a ResourceGroup as inventory object soon.

**Review by mortent — APPROVED** [review]

**Pull request merged**